### PR TITLE
Retrieve the Dbpedia ontology

### LINKED ISSUE #108: Retrieve the Dbpedia ontology

### PR DESCRIPTION
**Pull request by doogyb**
This PR introduces code which automatically scrapes the current dbpedia
ontology and saves it to a json format.

closes https://github.com/nestauk/dap_dv_backends/issues/108

### CODE REVIEW

File: src/node_modules/sparql/util.mjs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export const prefixes =
```
Comment (mindrones) [inline src/node_modules/sparql/util.mjs]: I'd move this file to `src/node_modules/dbpedia/util.mjs` as it's specific to it

File: src/node_modules/dbpedia/requests.mjs [hunk 2]
```diff
@@ -1,22 +1,47 @@
...
+import { prefixes } from 'sparql/util.mjs';
 
-/**
- * @function getEntityDetails
- * @description provides details such as imageURL and abstract for supplied DBpedia URIs
- * @param {String|String[]} input - a single DBpedia URI or a list of URIs.
- * @returns a list of entities for the supplied DBPedia URIs.
- */
-export const getEntityDetails = async input => {
+const sanitizeInput = input => {
```
Comment (mindrones) [inline src/node_modules/dbpedia/requests.mjs:7]: I'd move `const dbr = 'http://dbpedia.org/resource';` to `src/node_modules/dbpedia/util.mjs` and would reuse it everywhere

File: src/node_modules/dbpedia/requests.mjs [hunk 3]
```diff
@@ -1,22 +1,47 @@
...
+    SELECT * WHERE {
+        ${body}
+    }`;
+	return sparql;
+};
+
+const makeRequest = async sparql => {
+	const { results } = await query(sparql);
+	const values = _.map(results.bindings, row =>
+    	_.mapValues(row, contents => contents.value)
```
Comment (mindrones) [inline src/node_modules/dbpedia/requests.mjs]: `contents => contents.value` can be written as `getValue` from `@svizzle/utils`

File: src/bin/dbpedia/parseOntology.mjs [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import * as cheerio from 'cheerio';
+import * as _ from 'lamb';
+import { fetch } from 'undici';
+import { saveObj } from '@svizzle/file';
```
Comment (mindrones) [inline src/bin/dbpedia/parseOntology.mjs]: imports to be sorted

File: src/bin/dbpedia/parseOntology.mjs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+const ONTOLOGY_URL = 'http://mappings.dbpedia.org/server/ontology/classes/';
+const FILE_ONTOLOGY_JSON = 'data/dbpedia/data/ontology.json';
+let $;
```
Comment (mindrones) [inline src/bin/dbpedia/parseOntology.mjs]: (1) `$` is being used in `main` only, so I'd declare it there

File: src/bin/dbpedia/parseOntology.mjs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+		class: className,
+		children: _.filter(_.map(children, parseOntology), child => child)
+	};
+};
+
+const main = async () => {
+
+	const response = await fetch(ONTOLOGY_URL);
+	const html = await response.text();
+	$ = cheerio.load(html);
```
Comment (mindrones) [inline src/bin/dbpedia/parseOntology.mjs]: See (1)
```suggestion
	const $ = cheerio.load(html);
```

File: src/bin/dbpedia/parseOntology.mjs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import { saveObj } from '@svizzle/file';
+
+const ONTOLOGY_URL = 'http://mappings.dbpedia.org/server/ontology/classes/';
+const FILE_ONTOLOGY_JSON = 'data/dbpedia/data/ontology.json';
+let $;
+
+const range = n => Array.from(Array(n).keys());
+const repeat = (char, n) => _.reduce(range(n), (acc, __) => `${acc}${char}`, char);
+
+const onlyLatinCharacters = str => (/^[a-zA-Z:]+$/u).test(str);
```
Comment (mindrones) [inline src/bin/dbpedia/parseOntology.mjs]: for booleans we usually have a verb first, so this would become `hasLatinCharsOnly`

File: src/bin/dbpedia/parseOntology.mjs [hunk 8]
```diff
@@ -0,0 +1,57 @@
+import * as cheerio from 'cheerio';
+import * as _ from 'lamb';
+import { fetch } from 'undici';
+import { saveObj } from '@svizzle/file';
+
+const ONTOLOGY_URL = 'http://mappings.dbpedia.org/server/ontology/classes/';
+const FILE_ONTOLOGY_JSON = 'data/dbpedia/data/ontology.json';
+let $;
+
+const range = n => Array.from(Array(n).keys());
```
Comment (mindrones) [inline src/bin/dbpedia/parseOntology.mjs]: we can use `_.range` here

File: src/bin/dbpedia/parseOntology.mjs [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import * as cheerio from 'cheerio';
+import * as _ from 'lamb';
+import { fetch } from 'undici';
+import { saveObj } from '@svizzle/file';
+
+const ONTOLOGY_URL = 'http://mappings.dbpedia.org/server/ontology/classes/';
+const FILE_ONTOLOGY_JSON = 'data/dbpedia/data/ontology.json';
```
Comment (mindrones) [inline src/bin/dbpedia/parseOntology.mjs]: I guess we donot need the deeper `data` dir so this could be just
```suggestion
const FILE_ONTOLOGY_JSON = 'data/dbpedia/ontology.json';
```

File: src/bin/dbpedia/parseOntology.mjs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+		return { class: className };
+	}
+	return {
+		class: className,
+		children: _.filter(_.map(children, parseOntology), child => child)
+	};
+};
+
+const main = async () => {
+
```
Comment (mindrones) [inline src/bin/dbpedia/parseOntology.mjs:45]: extra blank line to be removed

File: src/node_modules/dbpedia/ontology.mjs [hunk 11]
```diff
@@ -0,0 +1,43 @@
+import { promises as fs } from 'fs';
+
+import * as _ from 'lamb';
+
+const FILE_ONTOLOGY_JSON = 'data/dbpedia/data/ontology.json';
+
+const dfs = (node, depth, { squash=false, fullURI=false } = {}) => {
```
Comment (mindrones) [inline src/node_modules/dbpedia/ontology.mjs]: I’d rename `dfs` to a human readable string

File: src/node_modules/dbpedia/ontology.mjs [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+	const className = fullURI
+		? `http://dbpedia.org/ontology/${node.class}`
```
Comment (mindrones) [inline src/node_modules/dbpedia/ontology.mjs:9]: I'd reuse `dbr` here and elsewhere

**Review by mindrones — CHANGES_REQUESTED** [review]
1st pass

File: src/node_modules/dbpedia/ontology.mjs (hunk 12 above)
Comment (mindrones) [reply src/node_modules/dbpedia/ontology.mjs:9]: Whoops, misread the URI base sorry, I see now that it's `dbr` but `dbo`. 

I'd have that value in a const for ease its detection later and facilitate reuse (possibly in `src/node_modules/sparql/util.mjs`)

File: package.json [hunk 13]
```diff
@@ -11,6 +11,7 @@
 		"@aws-sdk/signature-v4": "^3.49.0",
 		"@svizzle/file": "^0.12.0",
 		"@svizzle/utils": "^0.16.0",
+		"cheerio": "^1.0.0-rc.10",
```
Comment (mindrones) [inline package.json]: unused dep to be removed

File: package.json [hunk 14]
```diff
@@ -38,6 +39,7 @@
 		"rebuild": "npm run rebuildTestIndex && npm run rebuildEdgeCasesIndex",
 		"rebuildEdgeCasesIndex": "node src/bin/es/index.mjs delete edge-cases && sleep 1 && node src/bin/es/index.mjs create edge-cases && sleep 1 && node src/bin/es/document.mjs create edge-cases ./src/test/data/elastic_search_results/test_documents.json",
 		"rebuildTestIndex": "node src/bin/es/index.mjs delete test && sleep 1 && node src/bin/es/index.mjs create test --path src/test/conf/test_index_mapping.json && sleep 1 && node src/bin/es/index.mjs reindex arxiv_v6 test --ignore dbpedia_entities dbpedia_entities_metadata --max-docs 1000",
+		"parseOntology": "node src/bin/dbpedia/parseOntology.mjs",
```
Comment (mindrones) [inline package.json]: I'd rename `src/bin/dbpedia/parseOntology.mjs` to `src/bin/dbpedia/getOntology.mjs`

File: src/bin/dbpedia/README.md [hunk 15]
```diff
@@ -0,0 +1,8 @@
+## parseOntology.mjs
+
+Script to parse the DBpedia ontology. 
+
+Fetches the html from the DBpedia ontology page
+http://mappings.dbpedia.org/server/ontology/classes/
+and uses the `cheerio` library to recursively search the html to generate
+the hierarchal structure for the ontology.
```
Comment (mindrones) [inline src/bin/dbpedia/README.md]: to be updated after moving to fetching the ontology via Sparql

File: src/bin/dbpedia/getDBpediaOntology.sparql [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+  ?class_ a owl:Class .
+  ?class_ rdfs:isDefinedBy <http://dbpedia.org/ontology/> .
+  ?class_ rdfs:subClassOf ?parent .
+  {
+    ?parent rdfs:isDefinedBy <http://dbpedia.org/ontology/>
+  }
+  UNION
+  {
+    ?parent a owl:Class .
+    ?parent rdfs:label "Thing" .
```
Comment (mindrones) [inline src/bin/dbpedia/getDBpediaOntology.sparql]: I'd rename `parent` to `parentClass`

File: src/node_modules/dbpedia/requests.mjs [hunk 17]
```diff
@@ -54,3 +68,28 @@ export const getEntityDetails = async input => {
...
+	return _.mapValues(groups, group => {
+		const filtered = _.filter(group, item => classFilter.has(item.type));
+		return _.map(filtered, _.getKey('type'));
+	});
+};
+
+const main = async () => {
+	const results = await getClasses(['Earth', 'Moon'], 0);
+	console.log(results);
+};
```
Comment (mindrones) [inline src/node_modules/dbpedia/requests.mjs]: seems unused for now, I'd remove

File: src/node_modules/util/array.mjs [hunk 18]
```diff
@@ -11,4 +11,5 @@ const _batch = (arr, batchSize) => {
 };
 
 export const batch = _.pipe([_batch, _.filterWith(isNotNil)]);
+export const range = n => Array.from(Array(n).keys());
```
Comment (mindrones) [inline src/node_modules/util/array.mjs:17]: we can use `_.range` so this is not needed :)

File: src/node_modules/util/string.mjs [hunk 19]
```diff
@@ -11,3 +11,16 @@ export const dedent = _.pipe([
...
+		str.slice(sep.lastIndex, sep.exec(str).index)
+	);
+	const result = [...prefix, str.slice(sep.lastIndex)];
+	return result;
+};
+
+export const onlyLatinCharacters = str => (/^[a-zA-Z:]+$/u).test(str);
+export const hasNonAsciiCharacters = str => (/^[\u0000-\u007f]*$/u).test(str);
+export const repeat = (char, n) =>
+	_.reduce(range(n), (acc, __) => `${acc}${char}`, char);
```
Comment (mindrones) [inline src/node_modules/util/string.mjs]: seems unused (also we have `_.repeat`! :)

File: src/node_modules/dbpedia/ontology.mjs [hunk 20]
```diff
@@ -0,0 +1,45 @@
+import { promises as fs } from 'fs';
```
Comment (mindrones) [inline src/node_modules/dbpedia/ontology.mjs]: as `loadOntology` is only used by `getClasses` which is unused, I'd remove this file

File: src/test/ai_map.spec.mjs [hunk 21]
```diff
@@ -145,4 +145,5 @@ describe('aiMapDBpediaEntities', () => {
 	});
 });
 
-
+// how many entities have classes not present in the DBpedia Ontology
+// which entities have non DBO classes
```
Comment (mindrones) [inline src/test/ai_map.spec.mjs]: leftover

File: src/node_modules/util/string.mjs [hunk 22]
```diff
@@ -11,3 +11,16 @@ export const dedent = _.pipe([
...
 	trim
 ]);
+
+export const splitAtMost = (str, sep, n) => {
+	const prefix = _.map(range(n), () =>
+		str.slice(sep.lastIndex, sep.exec(str).index)
+	);
+	const result = [...prefix, str.slice(sep.lastIndex)];
+	return result;
+};
```
Comment (mindrones) [inline src/node_modules/util/string.mjs]: seems unused

**Review by mindrones — CHANGES_REQUESTED** [review]
pass 2

File: src/node_modules/util/string.mjs (hunk 22 above)
Comment (doogyb) [reply src/node_modules/util/string.mjs]: I might check this code into svizzle then? Could be handy to have in future.

File: src/node_modules/dbpedia/requests.mjs (hunk 17 above)
Comment (doogyb) [reply src/node_modules/dbpedia/requests.mjs]: This is the logic we spoke of that allows us to retrieve classes for entities. I just have to update the `loadOntology` functionality but otherwise this is the function we'll use in the subsequent PR to populate the `ai_map` entities' classes and run data quality tests on them in `ai_map.spec.mjs`.

File: src/bin/dbpedia/getDBpediaOntology.sparql [hunk 23]
```diff
@@ -0,0 +1,17 @@
+PREFIX dbo: <http://dbpedia.org/ontology/>
```
Comment (mindrones) [inline src/bin/dbpedia/getDBpediaOntology.sparql]: I'd rename to `src/bin/dbpedia/getOntology.sparql` to keep consistent with the script

File: src/bin/dbpedia/README.md [hunk 24]
```diff
@@ -0,0 +1,26 @@
+## parseOntology.mjs
+
```
Comment (mindrones) [inline src/bin/dbpedia/README.md]:
```suggestion
## getOntology.mjs

```

File: src/bin/dbpedia/README.md [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+Script to parse the DBpedia ontology. 
+
+Fetches DBpedia classes and associated parent classes using a SPARQL query
+to the DBPedia endpoint. We assure classes are from the DBpedia ontology using
+the following predicate object combination: 
+`rdfs:isDefinedBy <http://dbpedia.org/ontology/> .`,
+The full SPARQL query can be found [here](getDBpediaOntology.sparql).
```
Comment (mindrones) [inline src/bin/dbpedia/README.md]:
```suggestion
The full SPARQL query can be found [here](./getOntology.sparql).
```

File: package.json [hunk 26]
```diff
@@ -34,6 +34,7 @@
 	},
 	"scripts": {
 		"annotate": "cross-env NODE_ENV=production NODE_NO_WARNINGS=1 node src/bin/es/annotate.mjs --field textBody_abstract_article --name dbpedia_entities --index arxiv_v6 --spotlight http://localhost:2222/rest/annotate --page-size 100 --include-metadata",
+		"getDBpediaOntology": "node src/bin/dbpedia/parseOntology.mjs",
```
Comment (mindrones) [inline package.json]:
```suggestion
		"getDBpediaOntology": "node src/bin/dbpedia/getOntology.mjs",
```

File: package-lock.json [hunk 27]
```diff
@@ -16,6 +16,7 @@
 				"@aws-sdk/signature-v4": "^3.49.0",
```
Comment (mindrones) [inline package-lock.json]: we need to re-install and update this lock file as it still mentions `cheerio`

File: src/node_modules/dbpedia/requests.mjs (hunk 3 above)
Comment (mindrones) [reply src/node_modules/dbpedia/requests.mjs]: we should use `const values = _.map(results.bindings, _.mapValues(getValue));` here

File: src/node_modules/dbpedia/requests.mjs (hunk 17 above)
Comment (mindrones) [reply src/node_modules/dbpedia/requests.mjs]: I'd make a copy somewhere and remove it (also cleaning up the code this depends on) as it's not in scope in this PR :)

File: src/node_modules/util/string.mjs (hunk 22 above)
Comment (mindrones) [reply src/node_modules/util/string.mjs]: let's discuss in slack!

**Review by mindrones — CHANGES_REQUESTED** [review]
pass 3

**Review by mindrones — APPROVED** [review]
Great, GTM!

(Note: that I haven't reviewed the `parse` function in `src/bin/dbpedia/getOntology.mjs`, we'll refactor that later on)
